<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
  "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="content-type" content=
"text/html; charset=iso-8859-1" />
<meta name="KeyWords" content=
"transcription factor activities, TFA, ChIP, chromatin, Korbinian Strimmer" />
<meta name="Author" content="Korbinian Strimmer" />
<meta name="Description" content="Links to GGM literature" />
<title>Reconstructing True Transcription Factor Activities</title>
<link rel="stylesheet" type="text/css" href="../format.css" />
</head>
<body>
<script type="text/javascript">
//<![CDATA[
  <!--  
   if (top.location != self.location) {top.location = self.location}
  //-->
//]]>
</script>
<div id="page">
<p class="c2">Notes: Reconstructing True Transcription Factor
Activities</p>
<p>(back to <a href="index.html">index</a>)</p>
<p>&nbsp;</p>
<p class="c1">Overview:</p>
<p><a href=
"http://en.wikipedia.org/wiki/Transcription_factor">Transcription
factors</a> are subject to strong post-translational modifications.
This has the consequence that mRNA concentrations, as measured,
e.g., by conventional microarray experiments cannot serve as
reliable proxies for the transcription factor activities (TFAs).
However, it turns out that it is possible to <em>estimate</em> the
TFAs from microrarray data (i.e. to correct the expression data) by
combining them in a suitable fashion with external connectivity
information, such as from <a href=
"http://en.wikipedia.org/wiki/Chromatin_immunoprecipitation">ChIP</a>
experiments.</p>
<p>This approach has been pioneered by the group around <a href=
"http://www.seas.ucla.edu/~liaoj/">J. C. Liao</a> at UCLA who
proposed the approach of <a href=
"http://www.google.de/search?q=%22network+component+analysis">Network
Component Analysis</a> (NCA) in a series of papers:</p>
<ol>
<li>Y.-L. Yang et al., J. C. Liao 2005. Inferring yeast cell cycle
regulators and interactions using transcription factor activities.
BMC Genomics 6:90.</li>
<li>L. M. Tran et al., J. C. Liao. 2005. gNCA: A framework for
determining transcription factor activity based on transcriptome:
identifiability and numerical implementation. Metab. Engin.
7:128-141.</li>
<li>R. Boscolo, C. Sabatti, J. C. Liao., and V. P. Roychowdhury.
2005. Reconstructing hidden regulatory layers by network component
analysis: theory and applications. IEEE Trans. Comput. Biol.
Bioinf. in press</li>
<li>K. C. Kao et al., J. C. Liao. 2004. Transcriptome-based
determination of multiple transcription regulator activities in
Escherichia coli by using network component analysis. PNAS 101:
641-646.</li>
<li>J. C. Liao, et al. 2003. Network component analysis:
reconstruction of regulatory signals in biological systems. PNAS
100:15522-15527.</li>
</ol>
<p>&nbsp;</p>
<p class="c1">Statistical Approaches to NCA:</p>
<p>The original NCA algorithm to infer the TFAS is rather adhoc and
does not take account of stochasticity, so several groups have
attempted to cast the method into a more statistical framework:</p>
<ol>
<li>F. Gao, B. C. Foat, and H. J. Bussemaker. 2004. Defining
transcriptional networks trough integrative modeling of mRNA
expression and transcription factor binding data. BMC
Bioinformatics 5:31</li>
<li>C. Sabatti and G. James. 2006. Bayesian sparse hidden
components analysis for transcription regulation networks.
Bioinformatics, in press. (UCLA Statistics Preprint #414)</li>
<li>A.-L. Boulesteix and K. Strimmer. 2005. Predicting
transcription factor activities from combined analysis of
microarray and ChIP data: a partial least squares approach.
<a href="http://www.tbiomed.com/content/2/1/23/abstract">Theor.
Biol. Med. Model. <strong>2</strong>: 23</a>. (<a href=
"publications/journals/regulation2005.pdf">preprint</a>)</li>
</ol>
<p>Gao et al. (2004) suggest to use linear regression with
step-wise model selection. Sabatti and James (2005) offer a
Bayesian approach to NCA. Our suggestion (<a href=
"http://www.tbiomed.com/content/2/1/23/abstract">Boulesteix and
Strimmer 2005</a>) is to use partial least-squares regression to
solve the NCA problem and estimate the TFAs (please refer to the
paper for details and a comparison of the above approaches).</p>
<p>&nbsp;</p>
<p class="c1">Further Related References:</p>
<ol>
<li>Z. Li and C. Chan. 2004. Extracting novel information from gene
expression data. Trends in Biotech. 22:381-383.</li>
<li>D. F. Simola. 2004. Prediction of transcription factor gene
expression in Saccaromyces cerevisae. Technical report, Univ.
Pennsylvania.</li>
<li style="list-style: none">O. Alter and G. H. Golub. 2004.
Integrative analysis of genome-scale data by using pseudoinverse
projection predicts novel correlation between DNA replication and
RNA transcription. PNAS 101:16577-16852.</li>
</ol>
<p>&nbsp;</p>
<p>Please drop me <a href="../index.html">me</a> a line (<a href=
"mailto:korbinian.strimmer@lmu.de">korbinian.strimmer@lmu.de</a>)
for suggestions and comments.</p>
<p>&nbsp;</p>
<p class="c3"><em>Last modified:</em><br />
July 21, 2005<br /></p>
<p class="c3"><a href=
"http://validator.w3.org/check?uri=referer"><img src=
"../images/valid-xhtml11.png" alt="Valid XHTML 1.1" height="31"
width="88" /></a></p>
</div>
</body>
</html>
